### PR TITLE
Add support for Mailbox Gateway (ETG.8200)

### LINKED ISSUE #60: Add support for ETG.8200 - Mailbox Gateway
Mailbox Gateway enable external mailbox access to a master that support it. It is useful for configuration tool and maintenance tools. 
Note: the Mailbox Gateway standard is required to add support for the Master Diagnostic standard (ETG.1510).

### PR DESCRIPTION
**Pull request by Siviuze**
close #60 
related to #57 (comment flooding traces, still need a better resolution)

### CODE REVIEW

File: src/Mailbox.cc [hunk 1]
```diff
@@ -341,7 +360,8 @@ namespace kickcat
         : AbstractMessage(mailbox.recv_size)
         , mailbox_{mailbox}
     {
-
+        // Is this needed?
```
Comment (Siviuze) [inline src/Mailbox.cc:363]: To check before merge

**Comment by Siviuze** [comment]
It is required to test it on a full EtherCAT network with CoE slaves before merge.

File: include/kickcat/Gateway.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include "Bus.h"
+#include "Mailbox.h"
+
+namespace kickcat
+{
+    /// \brief Handle external mailbox requests as defined in ETG.8200
+    class Gateway
+    {
+    public:
+        /// \param socket       Socker to communicate with external networks
```
Comment (Rdk-T) [inline include/kickcat/Gateway.h:14]: Typo Socker

File: include/kickcat/Gateway.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        Gateway(std::shared_ptr<AbstractDiagSocket> socket, std::function<std::shared_ptr<GatewayMessage>(uint8_t const*, int32_t, uint16_t)> addMessage)
+            : socket_{socket}
+            , addMessage_{addMessage}
+            , pendingRequests_{}
+        { }
+        virtual ~Gateway() = default;
+
+        /// \brief   Fetch a request on the external network and process it if any
+        /// \details Try to get a request from the network (non blocking call). If there is data, check the request coherency then
+        ///          transfer it to the network while adding it to the pending requests
```
Comment (Rdk-T) [inline include/kickcat/Gateway.h:25]: transfer it to the bus instead of the network no ?

File: include/kickcat/Gateway.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        void fetchRequest();
+
+        /// \brief   Process pending requests
+        /// \details A pending request is a request that was successfully transfered to the EtherCAT bus, waiting for its completion.
+        ///          All requests completed an the bus are sent back to their respective requestors
```
Comment (Rdk-T) [inline include/kickcat/Gateway.h:30]: an the bus ?

File: src/Gateway.cc [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            return;
+        }
+
+        // -> Frame command type
+        if (header->type != EthercatType::MAILBOX)
+        {
+            DEBUG_PRINT("Frame command is not MAILBOX but %d", header->type);
+            return;
+        }
+
```
Comment (Rdk-T) [inline src/Gateway.cc:34]: There is no info in the ETG8200 about what the gateway is suppose to return to the client in case of ill-formatted requests ?

Comment (Siviuze) [reply src/Gateway.cc:34]: From 4.1
```
There is no error handling specified for the Mailbox Gateway functionality. A request to a non-existing
slave device can lead to no Response from the Master. The Device Configuration Tool has to handle
this e.g. with a timeout.
```

File: include/kickcat/Gateway.h (hunk 4 above)
Comment (Siviuze) [reply include/kickcat/Gateway.h:30]: `on`, maybe `by` is better?

File: include/kickcat/Gateway.h (hunk 3 above)
Comment (Siviuze) [reply include/kickcat/Gateway.h:25]: Right: this is misleading and I am referring to the EtherCAT bus

File: include/kickcat/protocol.h [hunk 6]
```diff
@@ -29,13 +47,22 @@ namespace kickcat
...
+    constexpr uint16_t ETH_ETHERCAT_TYPE = hton<uint16_t>(0x88A4);
     constexpr int32_t ETHERCAT_WKC_SIZE = 2;
     constexpr int32_t MAX_ETHERCAT_DATAGRAMS = 15;       // max EtherCAT datagrams per Ethernet frame
     constexpr uint16_t MAX_ETHERCAT_PAYLOAD_SIZE = 1486; // max EtherCAT payload size possible
 
+    // EtherCAT subprotocol
+    enum EthercatType : int16_t
+    {
+        ETHERCAT = 0x1,             // EtherCAT Data Link Protocol Data Unit (the real time protocol)
+        NETWORK_VARIABLES = 0x4,
```
Comment (Rdk-T) [inline include/kickcat/protocol.h:62]: what are network variables ?

File: include/kickcat/Gateway.h (hunk 4 above)
Comment (Rdk-T) [reply include/kickcat/Gateway.h]: i would go for `on`

Comment (Siviuze) [reply include/kickcat/Gateway.h]: Done

File: include/kickcat/AbstractDiagSocket.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#define KICKCAT_ABSTRACT_DIAG_SOCKET_H
+
+#include "Error.h"
+#include "Frame.h"
+
+namespace kickcat
+{
+    /// \brief   A socket interface for diagnostic interface (cf. ETG.8200)
+    /// \details This socket serves as an abstraction on classical socket (UDP, TCP, etc.)
+    ///          and shall be able to associate and identifier to income request and reply
```
[the remark this answers is not in the document]
Comment (Siviuze) [reply include/kickcat/AbstractDiagSocket.h:11]: done

File: include/kickcat/protocol.h (hunk 6 above)
Comment (Siviuze) [reply include/kickcat/protocol.h:62]: Another network protocol that seems to be designed to let multiple masters to communicates. I add this entry to have all the possibles types in the code.

File: src/Mailbox.cc (hunk 1 above)
Comment (Siviuze) [reply src/Mailbox.cc]: Tested successfully on an Elmo Gold (moving a motor and throwing CoE emergency message).

File: include/kickcat/AbstractDiagSocket.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// \brief Read data on the network (no protocol check!)
+        /// \param frame        Buffer where to store the received data
+        /// \param frame_size   Size of the buffer to store data to
+        /// \return A tuple with the actual received size (-1 in case of error) and an identifier to associate an answer
+        virtual std::tuple<int32_t, uint16_t> recv(uint8_t* frame, int32_t frame_size) = 0;
+
+        /// \brief Send data on the network (no protocol check!)
+        /// \param frame        Buffer to send on the network
+        /// \param frame_size   Size of th ebuffer to send on the network
```
Comment (Rdk-T) [inline include/kickcat/AbstractDiagSocket.h:29]: typo th ebuzzer

**Review by Rdk-T — APPROVED** [review]

**Pull request merged**